### ISSUE TITLE
Adopting different image sampling strategies

### ISSUE THREAD

**Issue by xiaoche-24** [ultralytics/yolov5#12670]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

If I want to sample images using different sampling strategies, should I modify the create_dataloader function as follows?
`def create_dataloader(
    path,
    imgsz,
    batch_size,
    stride,
    single_cls=False,
    hyp=None,
    augment=False,
    cache=False,
    pad=0.0,
    rect=False,
    rank=-1,
    workers=8,
    image_weights=False,
    quad=False,
    prefix="",
    shuffle=False,
    seed=0,
):
    if rect and shuffle:
        LOGGER.warning("WARNING ⚠️ --rect is incompatible with DataLoader shuffle, setting shuffle=False")
        shuffle = False
    with torch_distributed_zero_first(rank):  # init dataset *.cache only once if DDP
        dataset = LoadImagesAndLabels(
            path,
            imgsz,
            batch_size,
            augment=augment,  # augmentation
            hyp=hyp,  # hyperparameters
            rect=rect,  # rectangular batches
            cache_images=cache,
            single_cls=single_cls,
            stride=int(stride),
            pad=pad,
            image_weights=image_weights,
            prefix=prefix,
            rank=rank,
        )

    batch_size = min(batch_size, len(dataset))
    nd = torch.cuda.device_count()  # number of CUDA devices
    nw = min([os.cpu_count() // max(nd, 1), batch_size if batch_size > 1 else 0, workers])  # number of workers
    # sampler = None if rank == -1 else SmartDistributedSampler(dataset, shuffle=shuffle)
    sampler = DistributedRepeatFactorReSampler(dataset)        《《《《------------------------------------------
    loader = DataLoader if image_weights else InfiniteDataLoader  # only DataLoader allows for attribute updates
    generator = torch.Generator()
    generator.manual_seed(6148914691236517205 + seed + RANK)
    return loader(
        dataset,
        batch_size=batch_size,
        shuffle=shuffle and sampler is None,
        num_workers=nw,
        sampler=sampler,
        pin_memory=PIN_MEMORY,
        collate_fn=LoadImagesAndLabels.collate_fn4 if quad else LoadImagesAndLabels.collate_fn,
        worker_init_fn=seed_worker,
        generator=generator,
    ), dataset`

**Comment by glenn-jocher** [comment]
@xiaoche-24 hello! 👋

Yes, if you want to implement different image sampling strategies, modifying the `create_dataloader` function is a good starting point. Your proposed change, where you replace the default sampler with a custom `DistributedRepeatFactorReSampler`, is a valid approach to customize the sampling behavior.

Just ensure that your custom sampler (`DistributedRepeatFactorReSampler` in this case) adheres to the PyTorch `Sampler` interface and is compatible with the rest of the data loading pipeline. Also, remember to handle any necessary adjustments for distributed training if you're using that.

For further details on customizing your data loader, you might want to check out the documentation on data loading and augmentation at our Ultralytics Docs.

Happy coding, and thank you for contributing to the YOLOv5 community! 😊
